### ISSUE TITLE
Placeholders vs labels

### ISSUE THREAD

**Issue by danielnixon** [twbs/bootstrap#15610]
Re https://github.com/twbs/bootstrap/pull/15327, I don't think HTML5 `placeholder`s are enough.

Should

    <span class="input-group-addon" id="basic-addon1">@</span>
    <input type="text" class="form-control" placeholder="Username" aria-describedby="basic-addon1">

be more like

    <span class="input-group-addon" id="basic-addon1">@</span>
    <input type="text" class="form-control" placeholder="you@example.com" aria-label="Username" aria-describedby="basic-addon1">

or even

    <label for="username" class="sr-only">Username</label>
    <span class="input-group-addon" id="basic-addon1">@</span>
    <input type="text" id="username" class="form-control" placeholder="you@example.com" aria-describedby="basic-addon1">

? See e.g. [MDN](https://developer.mozilla.org/en/docs/Web/HTML/Element/Input#attr-placeholder) or [The HTML5 placeholder attribute is not a substitute for the label element](http://www.456bereastreet.com/archive/201204/the_html5_placeholder_attribute_is_not_a_substitute_for_the_label_element/) or [Placeholders are not Substitutes for Labels](http://www.pardot.com/faqs/forms/placeholders-and-labels/).

**Comment by cvrebert** [comment]
But the user cannot change the cents field, it's always zero cents, so that nitpick doesn't seem right.

**Comment by danielnixon** [comment]
D'oh, you're right.

**Comment by patrickhlauke** [comment]
In the first example, the input is not an email address, but a twitter user name. the field is prefixed with the visible "@". ``placeholder="Username"`` is correct in this context.
As the ``placeholder`` attribute is exposed to assistive technologies (in case other methods, like a proper ``<label>``, an ``aria-label``, ``aria-labelledby`` or ``title`` are not present) as an accessible name, it functions in exactly the same way in this case as the proposed additional ``aria-label``.
The major objections to using ``placeholder`` as a form of labelling are usability-related. In an ideal world devs would always provide a visible ``<label>``, but when they decide not to, the above pattern is appropriate (and works well for assistive technologies too).

**Comment by patrickhlauke** [comment]
Related though, perhaps it's worth adding an example that shows the use of an ``<input>`` with an ``.input-group-addon``, but also a proper ``<label>`` ... will slot that on my to-do list

**Comment by mdo** [comment]
Going to punt on this, though if you wanted to do a label variation for the docs @patrickhlauke, I'm not against it.
